### PR TITLE
Fix random state for dropout_layer_norm

### LINKED ISSUE #314: dropout_layer_norm isn't random
`dropout_add_ln_fwd` doesn't seem to change the random state, which causes consecutive calls to use an identical dropout mask. I traced the problem to the counter offset being zero (https://github.com/HazyResearch/flash-attention/blob/main/csrc/layer_norm/ln_api.cpp#L258-L263), not sure what the actual value should be.

The problem can be seen in the following script, which compares to a correct behavior from torch.dropout (I added `std::cout<<"COUNTER OFFSET "<<counter_offset<<std::endl;` in the C code)
```
import torch
import dropout_layer_norm

kwargs={"device":"cuda", "dtype":torch.float16}
d=4096
torch.cuda.init()
torch.cuda.manual_seed(0)
gen=torch.cuda.default_generators[torch.cuda.current_device()]
print("START",gen.get_state()[-8:].tolist())
x=torch.randn([d,d], **kwargs)
r=torch.randn([d,d], **kwargs)
gamma=torch.randn([d], **kwargs)
beta=torch.randn([d], **kwargs)

print("BEFORE DROPOUT LN",gen.get_state()[-8:].tolist())
y1, *rest = dropout_layer_norm.dropout_add_ln_fwd(
        x, r, gamma, beta, None, None, None, None, 0.1, 1e-5,
        1.0, 0, None, False, False
    )
print("AFTER DROPOUT LN 1",gen.get_state()[-8:].tolist())
y2, *rest=dropout_layer_norm.dropout_add_ln_fwd(
        x, r, gamma, beta, None, None, None, None, 0.1, 1e-5,
        1.0, 0, None, False, False
    )
print("AFTER DROPOUT LN 1",gen.get_state()[-8:].tolist())
print("MATCHING", torch.equal(y1,y2))
y3=torch.dropout(x, 0.1, True)
print("AFTER DROPOUT 1",gen.get_state()[-8:].tolist())
y4=torch.dropout(x, 0.1, True)
print("AFTER DROPOUT 2",gen.get_state()[-8:].tolist())
print("MATCHING", torch.equal(y3,y4))
```
```
START [0, 0, 0, 0, 0, 0, 0, 0]
BEFORE DROPOUT LN [160, 0, 0, 0, 0, 0, 0, 0]
COUNTER OFFSET 0
AFTER DROPOUT LN 1 [160, 0, 0, 0, 0, 0, 0, 0]
COUNTER OFFSET 0
AFTER DROPOUT LN 1 [160, 0, 0, 0, 0, 0, 0, 0]
MATCHING True
AFTER DROPOUT 1 [236, 0, 0, 0, 0, 0, 0, 0]
AFTER DROPOUT 2 [56, 1, 0, 0, 0, 0, 0, 0]
MATCHING False
```
I'm running version 72ad03e (except for the modification above) on a DGX-A100 with Cuda 12.1, but it shouldn't matter since the bug has little to do with cuda.

### PR DESCRIPTION
**Pull request by jlamypoirier**
Fixes #314 

See #314. Moving the `launcher` call ensures `params.rows` is defined and fixes the counter value, I think it shouldn't have any other effect.

Running the script
```


import torch
import dropout_layer_norm

kwargs={"device":"cuda", "dtype":torch.float16}
d=4096
torch.cuda.init()
torch.cuda.manual_seed(0)
gen=torch.cuda.default_generators[torch.cuda.current_device()]
print("START",gen.get_state()[-8:].tolist())
x=torch.randn([d,d], **kwargs)
r=torch.randn([d,d], **kwargs)
gamma=torch.randn([d], **kwargs)
beta=torch.randn([d], **kwargs)

print("BEFORE DROPOUT LN",gen.get_state()[-8:].tolist())
y1, *rest = dropout_layer_norm.dropout_add_ln_fwd(
        x, r, gamma, beta, None, None, None, None, 0.1, 1e-5,
        1.0, 0, None, False, False
    )
print("AFTER DROPOUT LN 1",gen.get_state()[-8:].tolist())
y2, *rest=dropout_layer_norm.dropout_add_ln_fwd(
        x, r, gamma, beta, None, None, None, None, 0.1, 1e-5,
        1.0, 0, None, False, False
    )
print("AFTER DROPOUT LN 1",gen.get_state()[-8:].tolist())
print("MATCHING", torch.equal(y1,y2))
y3=torch.dropout(x, 0.1, True)
print("AFTER DROPOUT 1",gen.get_state()[-8:].tolist())
y4=torch.dropout(x, 0.1, True)
print("AFTER DROPOUT 2",gen.get_state()[-8:].tolist())
print("MATCHING", torch.equal(y3,y4))
```

Before fix (random state unchanged, 2nd call identical; dropout for reference)
```
START [0, 0, 0, 0, 0, 0, 0, 0]
BEFORE DROPOUT LN [160, 0, 0, 0, 0, 0, 0, 0]
AFTER DROPOUT LN 1 [160, 0, 0, 0, 0, 0, 0, 0]
AFTER DROPOUT LN 1 [160, 0, 0, 0, 0, 0, 0, 0]
MATCHING True
AFTER DROPOUT 1 [236, 0, 0, 0, 0, 0, 0, 0]
AFTER DROPOUT 2 [56, 1, 0, 0, 0, 0, 0, 0]
MATCHING False
```

After fix (random state changed, 2nd call different
```
START [0, 0, 0, 0, 0, 0, 0, 0]
BEFORE DROPOUT LN [160, 0, 0, 0, 0, 0, 0, 0]
AFTER DROPOUT LN 1 [224, 1, 0, 0, 0, 0, 0, 0]
AFTER DROPOUT LN 2 [32, 3, 0, 0, 0, 0, 0, 0]
MATCHING False
AFTER DROPOUT 1 [108, 3, 0, 0, 0, 0, 0, 0]
AFTER DROPOUT 2 [184, 3, 0, 0, 0, 0, 0, 0]
MATCHING False
```

### CODE REVIEW

**Comment by sailordiary** [comment]
If I'm read it correctly, this seems like a major problem -- so why isn't this PR merged?

**Comment by tridao** [comment]
@sailordiary I'm traveling, will take a look later.
@jlamypoirier Thanks so much for the thorough investigation!
